larpandoracontent/LArThreeDReco/LArHitCreation/RANSACMethod.cc

```diff
@@ -0,0 +1,490 @@
+    const float dispFromFitEnd = (hitPosition - fitEnd).GetDotProduct(fitDirection);
+
+    return (dispFromFitEnd > 0.0 && std::abs(dispFromFitEnd) < threshold);
+}
+
+void projectedHitDisplacement(RANSACHit &hit, const ThreeDSlidingFitResult slidingFit)
+{
+    const CartesianVector hitPosition = hit.GetProtoHit().GetPosition3D();
+
+    CartesianVector projectedPosition(0.f, 0.f, 0.f);
+    CartesianVector projectedDirection(0.f, 0.f, 0.f);
+    const float rL(slidingFit.GetLongitudinalDisplacement(hitPosition));
+    const StatusCode positionStatusCode(slidingFit.GetGlobalFitPosition(rL, projectedPosition));
+    const StatusCode directionStatusCode(slidingFit.GetGlobalFitDirection(rL, projectedDirection));
+
+    if (positionStatusCode != STATUS_CODE_SUCCESS || directionStatusCode != STATUS_CODE_SUCCESS)
+        return;
+
+    const float displacement = (hitPosition - projectedPosition).GetCrossProduct(projectedDirection).GetMagnitude();
+
+    hit.SetDisplacement(displacement);
+}
+
+void hitDisplacement(RANSACHit &hit, const CartesianVector fitEnd, const CartesianVector fitDirection)
+{
+    const CartesianVector hitPosition = hit.GetProtoHit().GetPosition3D();
+    const float displacement = (hitPosition - fitEnd).GetCrossProduct(fitDirection).GetMagnitude();
+
+    hit.SetDisplacement(displacement);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void LArRANSACMethod::ExtendFit(
+    std::list<RANSACHit> &hitsToTestAgainst,
+    RANSACHitVector &hitsToUseForFit,
+    std::vector<RANSACHit> &hitsToAdd,
+    const float distanceToFitThreshold,
+    const ExtendDirection extendDirection
+)
+{
+    if (hitsToUseForFit.size() == 0)
+        return;
+
+    float distanceToEndThreshold = 10; // TODO: Config?
+
+    CartesianPointVector fitPoints;
+    for (auto hit : hitsToUseForFit)
+        fitPoints.push_back(hit.GetProtoHit().GetPosition3D());
+
+    const unsigned int layerWindow(100);
+    const float pitch(LArGeometryHelper::GetWireZPitch(this->GetPandora()));
+    const ThreeDSlidingFitResult slidingFit(&fitPoints, layerWindow, pitch);
+
+    CartesianVector fitDirection = slidingFit.GetGlobalMaxLayerDirection();
+    CartesianVector fitEnd = slidingFit.GetGlobalMaxLayerPosition();
+
+    if (extendDirection == ExtendDirection::Backward)
+    {
+        // TODO: This seems a bit iffy...does this work as I expect?
+        fitDirection = slidingFit.GetGlobalMinLayerDirection();
+        fitEnd = slidingFit.GetGlobalMinLayerPosition();
+        fitDirection = fitDirection * -1.0;
+    }
+
+    // INFO: Check if there is a detector gap, and if there is, extend the fit over it.
+    auto distanceToProjectOverGap = LArGeometryHelper::ProjectAcrossGap3D(this->GetPandora(), fitEnd, fitDirection, 2.0, 10);
+    if (distanceToProjectOverGap > 0)
+        distanceToEndThreshold += std::min(distanceToProjectOverGap, 50.0f);
```

Comment:
Has any tuning been done on this cap? If not, what's the general thinking behind 50?

Reply:
No tuning, it was just a value that felt right to me and Andy. 50 should be 50cm here, which felt about right for a steep track going into a gap (including a tolerance on either side, so it ends up smaller than 50 for the gap size).